File: packages/core/src/checks/least-privilege.ts
import { BaseCheck, SkillContext, CheckResult } from '../types';

export class LeastPrivilegeCheck implements BaseCheck {
  id = 'least-privilege';
  name = 'Least Privilege Audit';
  weight = 0.8;

  async run(context: SkillContext): Promise<CheckResult> {
    const dangerous = ['run_shell_command', 'write_file', 'delete_file'];
    const found = context.tools.filter(t => dangerous.includes(t));

    if (found.length === 0) {
      return {
        id: this.id,
        name: this.name,
        score: 10,
        level: 'Low',
        justification: 'No high-risk tools requested.'
      };
    }

    const keywords: Record<string, string[]> = {
      'run_shell_command': ['bash', 'shell', 'command', 'terminal', 'execute', 'cli', 'run'],
      'write_file': ['save', 'write', 'create', 'file', 'output', 'export'],
      'delete_file': ['remove', 'delete', 'unlink', 'cleanup', 'rm']
    };

    const prompt = context.systemPrompt.toLowerCase();
    const unjustified = found.filter(tool => {
      const toolKeywords = keywords[tool] || [];
      return !toolKeywords.some(k => prompt.includes(k));
    });

    if (unjustified.length > 0) {
      return {
        id: this.id,
        name: this.name,
        score: 5,
        level: 'High',
        justification: `High-risk tools requested without clear justification in instructions: ${unjustified.join(', ')}`
      };
    }

    return {
      id: this.id,
      name: this.name,
      score: 10,
      level: 'Low',
      justification: 'Requested tools are justified by the skill instructions.'
    };
  }
}

File: internal/checks/least_privilege.go
package checks

import (
	"context"
	"fmt"
	"strings"

	"github.com/albertowar/skillauditai/internal/behavioral"
	"github.com/albertowar/skillauditai/pkg/api"
)

type LeastPrivilegeCheck struct{}

func (c *LeastPrivilegeCheck) ID() string      { return "least-privilege" }
func (c *LeastPrivilegeCheck) Name() string    { return "Least Privilege Audit" }
func (c *LeastPrivilegeCheck) Weight() float64 { return 0.8 }

func (c *LeastPrivilegeCheck) Run(ctx context.Context, skill api.SkillContext, b *behavioral.Service) (api.CheckResult, error) {
	dangerous := []string{"run_shell_command", "write_file", "delete_file"}
	var found []string
	for _, t := range skill.Tools {
		lowerTool := strings.ToLower(t)
		for _, d := range dangerous {
			if lowerTool == d {
				found = append(found, t)
				break
			}
		}
	}

	if len(found) == 0 {
		return api.CheckResult{
			ID:            c.ID(),
			Name:          c.Name(),
			Score:         10,
			Level:         api.Low,
			Justification: "No high-risk tools requested.",
		}, nil
	}

	keywords := map[string][]string{
		"run_shell_command": {"bash", "shell", "command", "terminal", "execute", "cli", "run"},
		"write_file":        {"save", "write", "create", "file", "output", "export"},
		"delete_file":       {"remove", "delete", "unlink", "cleanup", "rm"},
	}

	prompt := strings.ToLower(skill.SystemPrompt)
	var unjustified []string

	for _, tool := range found {
		lowerTool := strings.ToLower(tool)
		toolKeywords := keywords[lowerTool]
		hasKeyword := false
		for _, k := range toolKeywords {
			if strings.Contains(prompt, k) {
				hasKeyword = true
				break
			}
		}
		if !hasKeyword {
			unjustified = append(unjustified, tool)
		}
	}

	if len(unjustified) > 0 {
		return api.CheckResult{
			ID:            c.ID(),
			Name:          c.Name(),
			Score:         5,
			Level:         api.High,
			Justification: fmt.Sprintf("High-risk tools requested without clear justification in instructions: %s", strings.Join(unjustified, ", ")),
		}, nil
	}

	return api.CheckResult{
		ID:            c.ID(),
		Name:          c.Name(),
		Score:         10,
		Level:         api.Low,
		Justification: "Requested tools are justified by the skill instructions.",
	}, nil
}
